test_mavros/include/tests/offboard_control.h

```diff
@@ -0,0 +1,508 @@
+} control_mode;
+
+typedef enum {
+	SQUARE,
+	CIRCLE,
+	EIGHT,
+	ELLIPSE
+} path_shape;
+
+class OffboardControl {
+public:
+	OffboardControl() :
+		nh_sp("~"),
+		mode_("position"),
+		shape_("square"),
+		local_pos_sp_pub(nh_sp.advertise<geometry_msgs::PoseStamped>("/mavros/setpoint_position/local", 10)),
+		vel_sp_pub(nh_sp.advertise<geometry_msgs::TwistStamped>("/mavros/setpoint_velocity/cmd_vel", 10)),
+		local_pos_sub(nh_sp.subscribe("/mavros/local_position/local", 10, &OffboardControl::local_pos_cb, this))
+	{ };
+
+	void init() {
+		/**
+		 * @brief Setpoint control mode selector
+		 *
+		 * Available modes:
+		 * - position
+		 * - velocity
+		 * - acceleration
+		 */
+		nh_sp.param<std::string>("mode", mode_, "position");
+
+		/**
+		 * @brief Setpoint path shape selector
+		 *
+		 * Available shapes:
+		 * - square
+		 * - circle
+		 * - eight
+		 * - ellipse (in 3D space)
+		 */
+		nh_sp.param<std::string>("shape", shape_, "square");
+
+		if (mode_ == "position")
+			mode = POSITION;
+		else if (mode_ == "velocity")
+			mode = VELOCITY;
+		else if (mode_ == "acceleration")
+			mode = ACCELERATION;
+		else {
+			ROS_ERROR_NAMED("sitl_test", "Control mode: wrong/unexistant control mode name %s", mode_.c_str());
+			return;
+		}
+
+		if (shape_ == "square")
+			shape = SQUARE;
+		else if (shape_ == "circle")
+			shape = CIRCLE;
+		else if (shape_ == "eight")
+			shape = EIGHT;
+		else if (shape_ == "ellipse")
+			shape = ELLIPSE;
+		else {
+			ROS_ERROR_NAMED("sitl_test", "Path shape: wrong/unexistant path shape name %s", shape_.c_str());
+			return;
+		}
+	}
+
+	/* -*- main routine -*- */
+
+	void spin(int argc, char *argv[]) {
+		init();
+		ros::Rate loop_rate(10);
+		ROS_INFO("SITL Test: Offboard control test running!");
+
+		if (mode == POSITION) {
+			ROS_INFO("Position control mode selected.");
+			if (shape == SQUARE) {
+				ROS_INFO("Test option: square-shaped path...");
+				square_path_motion(loop_rate, mode);
+			}
+			else if (shape == CIRCLE) {
+				ROS_INFO("Test option: circle-shaped path...");
+				circle_path_motion(loop_rate, mode);
+			}
+			else if (shape == EIGHT) {
+				ROS_INFO("Test option: eight-shaped path...");
+				eight_path_motion(loop_rate, mode);
+			}
+			else if (shape == ELLIPSE) {
+				ROS_INFO("Test option: ellipse-shaped path...");
+				ellipse_path_motion(loop_rate, mode);
+			}
+		}
+
+		else if (mode == VELOCITY) {
+			ROS_INFO("Velocity control mode selected.");
+			if (shape == SQUARE) {
+				ROS_INFO("Test option: square-shaped path...");
+				square_path_motion(loop_rate, mode);
+			}
+			else if (shape == CIRCLE) {
+				ROS_INFO("Test option: circle-shaped path...");
+				circle_path_motion(loop_rate, mode);
+			}
+			else if (shape == EIGHT) {
+				ROS_INFO("Test option: eight-shaped path...");
+				eight_path_motion(loop_rate, mode);
+			}
+			else if (shape == ELLIPSE) {
+				ROS_INFO("Test option: ellipse-shaped path...");
+				ellipse_path_motion(loop_rate, mode);
+			}
+		}
```

Comment:
I tried that once but it was failing and I thought it was because I was doing it. Looks like it wasn't because of it, cause I applied your suggestion now and worked.